feat: add oauth2 token exchange

code-asher opened the pull request:
Re-opened since the original got closed by stale bot and will not let me re-open after force-pushing!

`Rename token hashed_secret to refresh_hash` and later are new since that PR was last reviewed.  The main thing is that a secret prefix was added.

https://github.com/coder/coder/pull/11778

code-asher committed:
Add OAuth2 provider codes and tokens to database

code-asher committed:
Check indexes in dbmem

Might have no match, and a -1 will panic.

Emyrk committed:
minor dbauthz changes

Emyrk committed:
use model struct over rbac.ResourceXX

code-asher committed:
Rename token hashed_secret to refresh_hash

code-asher committed:
Format oauth2.sql

code-asher committed:
fixup! Rename token hashed_secret to refresh_hash

code-asher committed:
Add secret prefix column and query to get token

This will be used as an ID that we can prefix into the secrets
themselves.  This is so we can salt the hashed secrets.

The token query is for implementing the refresh flow.

code-asher commented on enterprise/coderd/oauth2.go:
```diff
@@ -208,10 +208,14 @@ func (api *API) postOAuth2ProviderAppSecret(rw http.ResponseWriter, r *http.Requ
 		})
 		return
 	}
+	// TODO: Currently unused.
+	prefix, _ := cryptorand.String(40)
```
Basically a stop-gap, this is replaced in the next PR.  I figure once this is approved I merged that one here, and then merge both into main together.

Emyrk commented on coderd/database/dbmem/dbmem.go:
```diff
@@ -1204,12 +1265,74 @@ func (q *FakeQuerier) DeleteOAuth2ProviderAppSecretByID(_ context.Context, id uu
 		if secret.ID == id {
 			q.oauth2ProviderAppSecrets[index] = q.oauth2ProviderAppSecrets[len(q.oauth2ProviderAppSecrets)-1]
 			q.oauth2ProviderAppSecrets = q.oauth2ProviderAppSecrets[:len(q.oauth2ProviderAppSecrets)-1]
+
+			// Cascade delete into tokens.
+			var keyIDsToDelete []string
+			var tokens []database.OAuth2ProviderAppToken
+			for _, token := range q.oauth2ProviderAppTokens {
+				if token.AppSecretID == id {
+					keyIDsToDelete = append(keyIDsToDelete, token.APIKeyID)
+				} else {
+					tokens = append(tokens, token)
+				}
+			}
+			q.oauth2ProviderAppTokens = tokens
+
+			// Delete from API keys.
+			var keys []database.APIKey
+			for _, key := range q.apiKeys {
+				if !slices.Contains(keyIDsToDelete, key.ID) {
+					keys = append(keys, key)
+				}
+			}
+			q.apiKeys = keys
+
 			return nil
```
nesting all this logic in the for loop feels a bit difficult to read. Maybe it's just me, but this feels a bit cleaner imo.

```golang
	index := slices.IndexFunc(q.oauth2ProviderAppSecrets, func(secret database.OAuth2ProviderAppSecret) bool {
		return secret.ID == id
	})

	if index < 0 {
		return sql.ErrNoRows
	}

	q.oauth2ProviderAppSecrets[index] = q.oauth2ProviderAppSecrets[len(q.oauth2ProviderAppSecrets)-1]
	q.oauth2ProviderAppSecrets = q.oauth2ProviderAppSecrets[:len(q.oauth2ProviderAppSecrets)-1]
	// ... rest of delete ...

```

code-asher replied:
You are right, this is much better.

code-asher replied:
I do this in a few places, going to update all of them along with using `DeleteFunc`!

code-asher commented on coderd/database/dbmem/dbmem.go:
```diff
@@ -1182,14 +1185,72 @@ func (q *FakeQuerier) DeleteOAuth2ProviderAppByID(_ context.Context, id uuid.UUI
+			}
+			q.oauth2ProviderAppTokens = tokens
+
+			// Delete from API keys.
+			var keys []database.APIKey
+			for _, key := range q.apiKeys {
+				if !slices.Contains(keyIDsToDelete, key.ID) {
+					keys = append(keys, key)
+				}
+			}
```
Oh snap TIL.  Will use.

code-asher committed:
Add oauth2 token exchange

code-asher committed:
Clean up dbmem funcs

code-asher committed:
Allow null on prefix after downgrade

Without this I think the tests would fail since they are not adding a
prefix and there is no default.
